web/source/keyboards/kmwkeyboards.ts

```diff
@@ -1155,7 +1161,7 @@ namespace com.keyman.keyboards {
         }
 
         if(first) {
-          this.keymanCloudRequest('',true);
+          return this.keymanCloudRequest('',true); // then?
         }
```

Comment:
@jahorton Should there be an else here to return a promise?

For when `this.languageList == null` and `first` is false.

Reply:
If my understanding of the original code here is correct...

```
if(first) {
  // KMW's language list is uninitialized AND is now needed
  // (because we just received a keyboard request based on language name)

  request_mapping_of_language_names_to_bcp47_codes()
} else {
  // the mapping request is still pending.
}
```

What you want:
- the `if(first)` query's `Promise` should be saved and stored; it (when implemented correctly) will guarantee that `this.languagesPending` has been initialized.  (That's what that specific, extra query does.)
- Use _that_ `Promise` to defer all incoming language-based queries, calling their `this.keymanCloudRequest()` calls from within their `then()` sections.

The "fun" then becomes figuring out the wiring needed to synchronously return a `Promise` based on an asynchronously-created internal `Promise`.

Reply:
I tried to use the nested promises https://github.com/keymanapp/keyman/pull/5260/commits/6899a526161243978fe653c189b651729c0d5338 from our 🦆 -session, but I still couldn't get it to behave (deferring languages before the language list is loaded)